File: archive/java/muse/data_structures/CircularlyLinkedList.java
package muse.data_structures;

import muse.data_structures.List;

public class CircularlyLinkedList<E> implements List<E> {
  private static class Node<E> {
    public E data;
    public Node<E> next = null;

    public Node(E element) {
      data = element;
    }
  }

  private Node<E> tail = null;
  private int length = 0;

  @Override
  public int size() {
    return length;
  }

  @Override
  public boolean isEmpty() {
    return length == 0;
  }

  @Override
  public E get(int index) {
    if (index < 0 || index >= length) {
      throw new RuntimeException("[PANIC - IndexOutOfBounds]");
    }

    Node<E> cursor = tail;

    if (index != length - 1) {
      for (int i = 0; i <= index; i++) {
        cursor = cursor.next;
      }
    }

    return cursor.data;
  }

  @Override
  public void set(int index, E element) {
    if (index < 0 || index >= length) {
      throw new RuntimeException("[PANIC - IndexOutOfBounds]");
    }

    Node<E> cursor = tail;

    if (index != length - 1) {
      for (int i = 0; i <= index; i++) {
        cursor = cursor.next;
      }
    }

    cursor.data = element;
  }

  @Override
  @SuppressWarnings("unchecked")
  public void insert(int index, E element) {
    if (index < 0 || index > length) {
      throw new RuntimeException("[PANIC - IndexOutOfBounds]");
    }

    Node<E> node = new Node<>(element);

    if (index == 0) {
      if (length == 0) {
        node.next = node;
        tail = node;
      } else {
        node.next = tail.next;
        tail.next = node;
      }
    } else if (index == length) {
      node.next = tail.next;
      tail.next = node;
      tail = node;
    } else {
      Node<E> cursor = tail;
      for (int i = 0, bound = index - 1; i <= bound; i++) {
        cursor = cursor.next;
      }
      node.next = cursor.next;
      cursor.next = node;
    }

    length++;
  }

  @Override
  public void remove(int index) {
    if (index < 0 || index >= length) {
      throw new RuntimeException("[PANIC - IndexOutOfBounds]");
    }

    Node<E> target;

    if (index == 0) {
      target = tail.next;
      if (length == 1) {
        tail = null;
      } else {
        tail.next = target.next;
      }
    } else {
      Node<E> cursor = tail;
      for (int i = 0, bound = index - 1; i <= bound; i++) {
        cursor = cursor.next;
      }
      target = cursor.next;
      cursor.next = target.next;
      if (index == length - 1) {
        tail = cursor;
      }
    }

    target.data = null;

    length--;
  }

  @Override
  public E front() {
    return get(0);
  }

  @Override
  public E back() {
    return get(length - 1);
  }

  @Override
  public void prepend(E element) {
    insert(0, element);
  }

  @Override
  public void append(E element) {
    insert(length, element);
  }

  @Override
  public void poll() {
    remove(0);
  }

  @Override
  public void eject() {
    remove(length - 1);
  }
}

File: archive/nodejs/muse/data_structures/CircularlyLinkedList.js
class Node {
  constructor(element) {
    this.data = element;
    this.next = null;
  }
}

export class CircularlyLinkedList {
  constructor() {
    this.tail = null;
    this.length = 0;
  }

  size() {
    return this.length;
  }

  isEmpty() {
    return this.length === 0;
  }

  get(index) {
    if (index < 0 || index >= this.length) {
      throw new Error('[PANIC - IndexOutOfBounds]');
    }

    let cursor = this.tail;

    if (index !== this.length - 1) {
      for (let i = 0; i <= index; i++) {
        cursor = cursor.next;
      }
    }

    return cursor.data;
  }

  set(index, element) {
    if (index < 0 || index >= this.length) {
      throw new Error('[PANIC - IndexOutOfBounds]');
    }

    let cursor = this.tail;

    if (index !== this.length - 1) {
      for (let i = 0; i <= index; i++) {
        cursor = cursor.next;
      }
    }

    cursor.data = element;
  }

  insert(index, element) {
    if (index < 0 || index > this.length) {
      throw new Error('[PANIC - IndexOutOfBounds]');
    }

    const node = new Node(element);

    switch (index) {
      case 0:
        if (this.length === 0) {
          node.next = node;
          this.tail = node;
        } else {
          node.next = this.tail.next;
          this.tail.next = node;
        }
        break;
      case this.length:
        node.next = this.tail.next;
        this.tail.next = node;
        this.tail = node;
        break;
      default: {
        let cursor = this.tail;
        for (let i = 0, bound = index - 1; i <= bound; i++) {
          cursor = cursor.next;
        }
        node.next = cursor.next;
        cursor.next = node;
      }
    }

    this.length++;
  }

  remove(index) {
    if (index < 0 || index >= this.length) {
      throw new Error('[PANIC - IndexOutOfBounds]');
    }

    let target = null;

    if (index === 0) {
      target = this.tail.next;
      if (this.length === 1) {
        this.tail = null;
      } else {
        this.tail.next = target.next;
      }
    } else {
      let cursor = this.tail;
      for (let i = 0, bound = index - 1; i <= bound; i++) {
        cursor = cursor.next;
      }
      target = cursor.next;
      cursor.next = target.next;
      if (index === this.length - 1) {
        this.tail = cursor;
      }
    }

    target.data = null;

    this.length--;
  }

  front() {
    return this.get(0);
  }

  back() {
    return this.get(this.length - 1);
  }

  prepend(element) {
    this.insert(0, element);
  }

  append(element) {
    this.insert(this.length, element);
  }

  poll() {
    this.remove(0);
  }

  eject() {
    this.remove(this.length - 1);
  }
}
